contents/blog/secrets-of-posthog-query-performance.md

```diff
@@ -0,0 +1,152 @@
+        HAVING max(is_deleted) = 0
+    )
+    GROUP BY distinct_id
+) AS pdi
+
+-- Table schema
+CREATE TABLE person_distinct_id
+(
+    distinct_id VARCHAR,
+    person_id UUID,
+    project_id Int64,
+    _sign Nullable(Int8),
+    is_deleted Nullable(Int8),
+    _timestamp DateTime
+) ENGINE = CollapsingMergeTree(_sign)
+```
+
+This JOIN is as complicated as it is due to a restriction from ClickHouse: updating data is expensive. In ClickHouse, updates require rewriting whole parts of the table instead of individual rows.
+
+Instead, the recommended approach is to use a ReplacingMergeTree or CollapsingMergeTree [TODO LINKS] table engine and handle updating logic at query-time.
+
+During data ingestion, when a given `distinct_id` had its `person_id` changed, PostHog emits a row with `is_deleted=1` for the old `person_id` and a new row with `is_deleted=0`. The above query would then resolve the `distinct_id` => `person_id` mapping at query time.
+
+However, in practice, this query took too much memory and was slow due to needing a subquery to aggregate data correctly. It also had subtle issues with using timestamps for versioning, causing issues.
+
+After noticing the problem, we realised we didn't need to actually emit rows with `is_deleted=0` to behave correctly, and could move to an alternative schema, which can be queried as follows:
+
+```sql
+JOIN (
+    SELECT distinct_id, argMax(person_id, version) as person_id
+    FROM person_distinct_id2
+    WHERE project_id = 123
+    GROUP BY distinct_id
+    HAVING argMax(is_deleted, version) = 0
+) AS pdi
+```
+
+For PostHog users with over 10 million visitors, this sped up queries previously bottlenecked on this JOIN by up to 10x.
+
+### Sorting events for a 23% win
+
+While debugging other performance issues, one question we kept asking is whether our data is laid out optimally for performance.
+
+PostHog uses a [ClickHouse MergeTree](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/mergetree/) table engine to store event data on disk. MergeTree tables can have an `ORDER BY` clause, which is then used by ClickHouse to store the data in a sorted format on disk and to create a sparse index of the data. The sparse index can then be used to skip reading data during queries.
+
+For PostHog our `ORDER BY` clause originally looked something like follows:
+
+```sql
+ORDER BY (project_id, toDate(timestamp), cityHash64(distinct_id), cityHash64(uuid))
+```
+
+Let’s consider this simplified query counting the number of users who did pageviews in a time range.
+
+```sql
+SELECT count(DISTINCT person_id)
+FROM events
+WHERE project_id = 2
+    AND event = '$pageview'
+    AND timestamp >= '2022-03-01'
+    AND timestamp < '2022-04-01'
+```
+
+When executing this query, ClickHouse is able to leverage data being sorted and the sparse index to skip reading most of data from disk. In this case, events from other projects and organizations and events from months other than March.
```

Comment:
the data -> rows change is just to avoid repetition, again just a nit
```suggestion
When executing this query, ClickHouse is able to leverage the sparse index and the sorted data to skip reading most of the rows from disk. In this case, events from other projects and organizations, and events from months other than March.
```

Reply:
This suggestion is slightly inaccurate as clickhouse reads columns not rows - will adjust the wording a bit differently.